File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    vector<vector<int>> memo;
    bool helper(vector<int>& nums,int i,int target)
    {
        if (target==0) return true;
        if (i>=nums.size() || target<0)
        {
            return false;
        }
        if(memo[i][target]!=-1){
            return memo[i][target];
        }
        return memo[i][target] = helper(nums,i+1,target-nums[i]) || helper(nums,i+1,target);
    }
    bool canPartition(vector<int>& nums) {
        
        int sum = 0;
        for(auto i: nums)
        {
            sum+=i;
        }
        if(sum%2!=0)return false;
        int target = sum/2;
        memo.resize(nums.size(),vector<int>(target+1,-1));
        return helper(nums,0,target);
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        dictt = {}
        summ = sum(nums)
        if summ%2!=0:
            return False
        def recursion(i,target):
            if target==0:
                return True
            if i>=len(nums):
                return False
            
            if (i,target) in dictt:
                return dictt[(i,target)]
            
            found = recursion(i+1,target-nums[i]) or recursion(i+1,target)
            dictt[(i,target)] = found
            return found
        
        return recursion(0,summ//2)
            
